app/src/main/java/com/google/android/apps/watchme/StreamerActivity.java

```diff
@@ -160,10 +168,80 @@ private void startStreaming() {
         wakeLock.acquire();
 
         if (!streamerService.isStreaming()) {
-            streamerService.startStreaming(rtmpUrl);
+
+            String cameraPermission = Manifest.permission.CAMERA;
+            String microphonePermission = Manifest.permission.RECORD_AUDIO;
+            int hasCamPermission = checkSelfPermission(cameraPermission);
+            int hasMicPermission = checkSelfPermission(microphonePermission);
+            List<String> permissions = new ArrayList<String>();
+            if (hasCamPermission != PackageManager.PERMISSION_GRANTED) {
+                permissions.add(cameraPermission);
+                if (ActivityCompat.shouldShowRequestPermissionRationale(this,
+                        Manifest.permission.CAMERA)) {
+                    // Provide rationale in Snackbar to request permission
+                    Snackbar.make(preview, R.string.permission_camera_rationale,
+                            Snackbar.LENGTH_INDEFINITE).show();
+                } else {
```

Comment:
I think this flow using the 'else' here is a little odd. For example, consider the new user scenario (no perms granted yet). Before we even request for permissions in app for the first time (at line 205), we already show a "turn on permissions in settings" toast (line 186), since 'shouldShowRequestPermissionRationale' is false initially.